### PR TITLE
Зайцев Артем. Задача 3. Вариант 25. Построение выпуклой оболочки. Проход Джарвиса.

### PR DESCRIPTION
**Pull request by aortemon**
## Jarvis march
1. Find the point, which is definitely is a part of the convex hull. It is accepted to choose the rightmost of the points having the minimal ordinate.
2. Among the others points it is necessary to find the most counterclockwise one. It means that slack vector formed by such a point and previously found start point has the minimal possible angle with all the other vectors that can be formed by other points of the set and selected the most counterclockwise one.
3. Repeat until the next point is the start point selected in p. 1.

## Sequential
1. Find the start point of the set, name it `start`.
2. Suggest that the next point of the given set is the next vertex of convex hull. Name it `next`.
5. For all the points of the set that have never been used before in this loop (name it `supposed`), check is triplet `(start, supposed, next)` the most counterclockwise among the others.
6. If it does, let `next` be `supposed`. It's the next vertex of convex hull.
7. Let `prev` be `next`, let `next` be the next point after actual `next`, and repeat 2-5 while points `next` and `start` are differentent.

## Parallel
The same as ***Sequentional***, but checking of triplet being the most counterclockwise is performed on several procs.

## struct Point
The struct defined in `seq/.../include/point.hpp` provides memebers `T x, y`, which represents the coords of the Point. `operator*(const Point& other)` represents vector product of vectors `(x, y)` and `(other.x, other.y)`. It's needed to simplify and shorten the solution. Also `b.between(const Point& a, const Point& c)` method is defined to check whether point `b` belongs to `ac` segment.  To be serializable by `boost` library and be sendable and recieveable by `boost::mpi`, `serialize(Archieve& a, const unsigned int version)` is overloaded for `Point<T>` as specified in `boost` docs.

### CODE REVIEW

File: tasks/mpi/zaitsev_a_jarvis/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+#include <vector>
+
+#include "boost/mpi/communicator.hpp"
+#include "mpi/zaitsev_a_jarvis/include/ops_mpi.hpp"
+#include "seq/zaitsev_a_jarvis/include/point.hpp"
+
+using Params =
+    std::tuple<std::vector<zaitsev_a_jarvis_seq::Point<double>>, std::vector<zaitsev_a_jarvis_seq::Point<double>>>;
+
```
Comment (Sean0450) [inline tasks/mpi/zaitsev_a_jarvis/func_tests/main.cpp:11]: Please, consider the cases when the points are collinear and have negative coordinates

Comment (aortemon) [reply tasks/mpi/zaitsev_a_jarvis/func_tests/main.cpp:11]: Added

File: tasks/mpi/zaitsev_a_jarvis/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      {{12, -11}, {15, 13}, {-8, 10}, {-13, 8}, {-11, -3}, {-9, -5}}
+    ),
+    //Common case with fractional case
+    Params(
+      {{0.4, 0.5}, {0.1, 0.2}, {0.2, -0.2}, {0.1, -0.5}, {-0.1, -0.1}, {-0.4, 0.4}, {-0.7, 0.6}, {-0.6, -0.4}, {-0.6, -0.7}},
+      {{-0.6, -0.7}, {0.1, -0.5}, {0.2, -0.2}, {0.4, 0.5}, {-0.7, 0.6}}
+    )
+  )
+);
+
```
[the remark this answers is not in the document]
Comment (aortemon) [reply tasks/mpi/zaitsev_a_jarvis/func_tests/main.cpp:113]: Thanks for review, your request helped me to find the lock in my code. Tests were added

[the remark this answers is not in the document]
Comment (aortemon) [reply tasks/mpi/zaitsev_a_jarvis/func_tests/main.cpp:113]: Coverage is increased up to 100%

image
image

**Review by PolinaSidorina — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

File: tasks/mpi/zaitsev_a_jarvis/include/ops_mpi.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+#include <boost/serialization/vector.hpp>
+#include <memory>
+#include <new>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/zaitsev_a_jarvis/include/point.hpp"
+
+#define EPS 1e-3
```
Comment (allnes) [inline tasks/mpi/zaitsev_a_jarvis/include/ops_mpi.hpp:16]: for this task its big error

Comment (aortemon) [reply tasks/mpi/zaitsev_a_jarvis/include/ops_mpi.hpp]: Changed

**Pull request merged**